src/section/section_proof_chain.rs

```diff
@@ -39,258 +38,209 @@ impl SectionProofBlock {
+    /// Creates new chain consisting of only one block.
+    pub fn new(first: SectionKeyInfo) -> Self {
         Self {
-            version: key_info.version,
-            prefix: key_info.prefix,
-            blocks: Vec::new(),
+            head: first,
+            tail: Vec::new(),
         }
     }
 
-    pub fn last_prefix_version(&self) -> (&Prefix<XorName>, u64) {
-        self.blocks
-            .last()
-            .map(|block| (block.prefix(), block.version()))
-            .unwrap_or((&self.prefix, self.version))
+    pub(crate) fn push(&mut self, block: SectionProofBlock) {
+        if !validate_next_block(self.last_key_info(), &block) {
+            log_or_panic!(
+                log::Level::Error,
+                "Invalid next block: {:?} -> {:?}",
+                self.last_key_info(),
+                block
+            );
+            return;
+        }
+
+        self.tail.push(block)
     }
 
-    #[cfg(all(test, feature = "mock"))]
-    pub fn all_prefix_version(&self) -> impl DoubleEndedIterator<Item = (&Prefix<XorName>, u64)> {
-        iter::once((&self.prefix, self.version)).chain(
-            self.blocks
-                .iter()
-                .map(|block| (block.prefix(), block.version())),
-        )
+    #[cfg(test)]
+    pub fn self_validate(&self) -> bool {
+        let mut current = &self.head;
+        for block in &self.tail {
+            if !validate_next_block(current, block) {
+                return false;
+            }
+
+            current = block.key_info();
+        }
+        true
     }
 
-    pub fn last_key_info(&self) -> Option<&SectionKeyInfo> {
-        self.blocks.last().map(|block| block.key_info())
+    pub(crate) fn first_key_info(&self) -> &SectionKeyInfo {
+        &self.head
     }
 
-    fn last_trusted_key_info<'a>(
-        &'a self,
-        last_trusted: &'a SectionKeyInfo,
-    ) -> Option<&'a SectionKeyInfo> {
-        let block_offset = last_trusted.version().saturating_sub(self.version) as usize;
+    pub(crate) fn last_key_info(&self) -> &SectionKeyInfo {
+        self.tail
+            .last()
+            .map(|block| block.key_info())
+            .unwrap_or(&self.head)
+    }
 
-        if block_offset == 0 {
-            if last_trusted.version() != self.version || last_trusted.prefix() != &self.prefix {
-                return None;
-            }
-        } else if let Some(block) = self.blocks.get(block_offset - 1) {
-            if block.key_info() != last_trusted {
-                return None;
-            }
-        } else {
-            // Root of trust not found
-            return None;
+    #[cfg(all(test, feature = "mock"))]
+    pub fn key_infos(&self) -> impl DoubleEndedIterator<Item = &SectionKeyInfo> {
+        iter::once(&self.head).chain(self.tail.iter().map(|block| block.key_info()))
+    }
+
+    /// Returns a slice of this chain starting at the given index.
+    pub(crate) fn slice_from(&self, first_index: usize) -> Self {
```

Comment:
why not construct a vector from the head & tail, then utilize the standard `split_off` function?
so the belowing un-explicit code block can then be avoided?

Reply:
Sounds like a good idea. Will give it a try.

Reply:
Ok, done. Didn't use `split_off`, but still simplified the impl.